src/Orchard/Mvc/MvcModule.cs

```diff
@@ -24,37 +27,25 @@ public class MvcModule : Module {
             moduleBuilder.Register(UrlHelperFactory).As<UrlHelper>().InstancePerDependency();
         }
 
-        private static bool IsRequestValid() {
-            if (HttpContext.Current == null)
-                return false;
-
-            try {
-                // The "Request" property throws at application startup on IIS integrated pipeline mode.
-                var req = HttpContext.Current.Request;
-            }
-            catch (Exception ex) {
-                if (ex.IsFatal()) {
-                    throw;
-                }
-
-                return false;
-            }
-
-            return true;
-        }
-
         static HttpContextBase HttpContextBaseFactory(IComponentContext context) {
-            if (IsRequestValid()) {
-                return new HttpContextWrapper(HttpContext.Current);
-            }
+
+            var httpContext = context.Resolve<IHttpContextAccessor>().Current();
+            if (httpContext != null)
+                return httpContext;
 
             var siteService = context.Resolve<ISiteService>();
+            var prefix = context.Resolve<ShellSettings>().RequestUrlPrefix;
 
             // Wrapping the code accessing the SiteSettings in a function that will be executed later (in HttpContextPlaceholder),
             // so that the RequestContext will have been established when the time comes to actually load the site settings,
             // which requires activating the Site content item, which in turn requires a UrlHelper, which in turn requires a RequestContext,
             // thus preventing a StackOverflowException.
-            var baseUrl = new Func<string>(() => siteService.GetSiteSettings().BaseUrl);
+            var baseUrl = new Func<string>(() => {
+                var url = siteService.GetSiteSettings().BaseUrl;
+                return !string.IsNullOrWhiteSpace(url) ? !string.IsNullOrWhiteSpace(prefix) ?

```

Comment:
@sebastienros just tried it again in the context of a tenant without hostnames (because in local environment) and a prefix = "alpha", then the baseUrl return the same as the default tenant. So, if we need a tenant valid url, we need the prefix, but maybe useless in a background task, I don't know, it's as you want.

@sfmskywalker you're right, I've seen this kind of loop, e.g when a factory needs itself... That's why I've created the `HttpContextPlaceholder.IsResolved` property (I use it in one factory) that checks the `baseUrl.IsValueCreated`...

 In my scenario it's quite different because the first resolution of the `HttpContextBase` is initiated by the workcontext scope creation, then carried through the CallContext and retrieved by the `HttpContextAccessor` that is now used by the `HttpContextBase` factory. So, when the workcontext scope is created, the factory only return what is found in the `CallContext`...

>Request object has no valid URL (even if it's a fake one) kittens will die a horrible death

Here I understand we need the baseUrl and a fake one is not enough. Do you mean as @sebastienros that fallback to a **fake one is useless**? And in case of a tenant do you think that the **prefix** is important?

 For infos, in my tests I have no problem to use a fake URL if I add the `ApplicationVirtualPath`.

Let me know your decision


Best